### PR TITLE
improve error handling in ws.decodeArrayBuffer and WebAudio backend

### LINKED ISSUE #1473: Failed to execute 'stop' on 'AudioScheduledSourceNode': cannot call stop without calling start first.
### Wavesurfer.js version(s):
wavesurfer.js 2.0.6 (Fri Jun 15 2018 10:57:58 GMT+0200 (CEST))

### Browser version(s):
Google Chrome: 69.0.3497.100 (Official Build) (64-bit)

### Code that causes the issue:

        /**
         * Pauses the loaded audio.
         */

    }, {
        key: 'pause',
        value: function pause() {
            this.scheduledPause = null;

            this.startPosition += this.getPlayedTime();
            this.source && this.source.stop(0);

            this.setState(PAUSED);

            this.fireEvent('pause');
        }

-A check is needed before executing this.source && this.source.stop(0);, because due to timing problems it might happen that the audio is not started when the stop is called. 

###
To reproduce the issue, you need to load new audio files, then quickly start and stop them.

### LINKED ISSUE #2279: TypeError: Cannot read property 'decodeArrayBuffer' of null
### Wavesurfer.js version(s):

5.0.1

### Browser and operating system version(s):

Nw.js / Electron

### Code needed to reproduce the issue:

```js
		fs.readFile(file.replace('file://', ''), (err, buffer) => {
			if (err) {
				console.error(err)
				return
			}
			const blob = new window.Blob([new Uint8Array(buffer)])
			this.wavesurfer.loadBlob(blob)
		})
```

### Use behaviour needed to reproduce the issue:

I am using WaveSurfer wrapped in a React component, I have a list of sounds and I can navigate through this list with keyboard shortcuts, when I go very fast through the list, the situation occurs that the wavesurfer's backend is gone and WaveSurfer crashes with: `TypeError: Cannot read property 'decodeArrayBuffer' of null`

 This happens over here: https://github.com/katspaugh/wavesurfer.js/blob/4360968534076ee7ed834e5c0c4ad48279f133a9/src/wavesurfer.js#L1542

I think we are missing a "this.isDestroyed" check over here: https://github.com/katspaugh/wavesurfer.js/blob/4360968534076ee7ed834e5c0c4ad48279f133a9/src/wavesurfer.js#L1326

Because loadBlob is using the FileReader which is asynchronous: https://github.com/katspaugh/wavesurfer.js/blob/4360968534076ee7ed834e5c0c4ad48279f133a9/src/wavesurfer.js#L1358

Here are some screenshots from me walking up the stack:

Screenshot 2021-05-26 at 15 46 19
Screenshot 2021-05-26 at 15 46 26
Screenshot 2021-05-26 at 15 46 32

### PR DESCRIPTION
**Pull request by vespakoen**
Fixes

`Failed to execute 'stop' on 'AudioScheduledSourceNode'`

and

`Cannot read property 'decodeArrayBuffer' of null`

(both timing / state issues)

### Short description of changes:

Adds the necessary checks to stop errors from happening.

### Breaking in the external API:

none

### Breaking changes in the internal API:

none

### Todos/Notes:

none

### Related Issues and other PRs:

fixes #2279
fixes #1473

### CODE REVIEW

File: src/webaudio.js [hunk 1]
```diff
@@ -702,7 +702,11 @@ export default class WebAudio extends util.Observer {
         this.scheduledPause = null;
 
         this.startPosition += this.getPlayedTime();
-        this.source && this.source.stop(0);
+        try {
+            this.source && this.source.stop(0);
+        } catch (err) {
+            // no-op, see https://github.com/katspaugh/wavesurfer.js/issues/1473
```
Comment (thijstriemstra) [inline src/webaudio.js:708]: can you add a little more info so you only have to visit the ticket if you need all details

**Review by thijstriemstra — CHANGES_REQUESTED** [review]
thanks! changes make sense. I preferred not to add a isDestroyed check everywhere but for this method I certainly don't mind.

can you also add a changelog entry? a new section with `x.x.x (unreleased)` is probably needed.

**Comment by thijstriemstra** [comment]
Also; bonus points for adding a unit test that would fail without this change.

**Comment by vespakoen** [comment]
I added the items to the changelog and a (hopefully) better comment.

I tried adding tests for these but it's pretty hard, since I don't know how I'd verify that an error threw.

I guess the test case would be something like:

```

    /** @test {WaveSurfer#destroy} */
    it('destroy while loading', function(done) {
        manualDestroy = true;

        wavesurfer.once('destroy', function() {
            TestHelpers.removeElement(element);
            done();
        });
        wavesurfer.load(TestHelpers.EXAMPLE_FILE_PATH);
        wavesurfer.destroy();

        // EXPECT NO THROWN ERRORS SOMEHOW?
    });
```

For the other issue it seems to happen when quickly starting / stopping a sound, perhaps something like:

```
    it('can quickly toggle', function() {
        wavesurfer.load(TestHelpers.EXAMPLE_FILE_PATH);
        wavesurfer.once('ready', () => {
            wavesurfer.play();
            wavesurfer.pause();
            wavesurfer.play();
            wavesurfer.pause();
            wavesurfer.play();
            wavesurfer.pause();
        });
    });
```

But it probably won't work all the time and is not elegant.

Basically, these are edge cases that happen in a rare situation that is hard to re-create in a test.

**Review by thijstriemstra — APPROVED** [review]
thanks!

**Comment by vespakoen** [comment]
Thank you too!

@thijstriemstra Would it be possible to do a minor release for this change? I'd like to include it in an update to one of my projects tomorrow.

If not, that's also fine ;)
